src/sonymn_int.cpp

```diff
@@ -795,4 +796,42 @@ namespace Exiv2 {
         return tagInfoCs2_;
     }
 
+    //! Sony Tag 9402 Sony2Fp (FocusPosition)
+    const TagInfo SonyMakerNote::tagInfoFp_[] = {
+        TagInfo(  0x04, "AmbientTemperature", N_("Ambient Temperature"), N_("Ambient Temperature"), sony2FpId, makerTags,   signedByte, 1, printValue),
+        TagInfo(  0x16, "FocusMode"         , N_("Focus Mode")         , N_("Focus Mode")         , sony2FpId, makerTags, unsignedByte, 1, printValue),
+        TagInfo(  0x17, "AFAreaMode"        , N_("AF Area Mode")       , N_("AF Area Mode")       , sony2FpId, makerTags, unsignedByte, 1, printValue),
+        TagInfo(  0x2d, "FocusPosition2"    , N_("Focus Position 2")   , N_("Focus Position 2")   , sony2FpId, makerTags, unsignedByte, 1, printValue),
+        // End of list marker
+        TagInfo(0xffff, "(Unknownsony2FpTag)", "(Unknownsony2FpTag)"   , "(Unknownsony2FpTag)"    , sony2FpId, makerTags, unsignedByte, 1, printValue)
+    };
+
+    const TagInfo* SonyMakerNote::tagListFp()
+    {
+        return tagInfoFp_;
+    }
+
+    DataBuf sonyFpCrypt(uint16_t /* tag */, const byte* bytes, uint32_t size, TiffComponent* const /*object*/)
+    {
+    	DataBuf b(bytes,size); // copy the data
+
+        // initialize the code table
+        byte  code[256];
+        for ( uint32_t i = 0 ; i < 249 ; i++ ) {
+            code[(i * i * i) % 249] = i ;
+        }
+        for ( uint32_t i = 249 ; i < 256 ; i++ ) {
+            code[i] = i;
+        }
+
+        // code byte-by-byte
+        uint32_t i = 0;
```

Comment:
[suggestion] I think that by replacing the while-loop by a for-loop the code could be reduced 2 lines.

Reply:
You're right.  I've changed it.

I still have the puzzle of encoding this data which I believe is causing the .exv file to be incorrect.  I will investigate that.  However, let's get this PR into v0.27.2 because it's good for reading the data which is the priority for @cryptomilk

I believe sonyFpCrypt() is a symmetric encrypt/decrypt.  The structure `ArrayCfg` has an crypt entry.  However it's not clear how the function knows if it's encrypting or decrypting and I don't want to change the API.

Reply:
> I believe sonyFpCrypt() is a symmetric encrypt/decrypt

In the ExifTool Decipher() routine, the $encipher parameter is a flag which is true to reverse the direction of the cipher (ie. encipher instead of decipher the data).  When you had this line of code:

`code[i] = (i * i * i) % 249 ;`

you were enciphering.  When you switched it to this:

`code[(i * i * i) % 249] =  i;`

you were deciphering.

Reply:
That's very useful, Phil.  There is a bug in my implementation when I write those tags.  I'm sure it's because writeMetadata() does not cipher the data.  I will try to fix that now with this insight.

Reply:
**YES**.  Thanks that's fixed the write in TiffEncoder::visitBinaryArray().  Muchy thanky.

I'll update the explanation I posted this morning to add a parameter 'decipher/encipher' to the utility sony_decipher (which will be renamed sony_cipher).  Very pleased about this.  Thank You very much for you help.